userapi/src/main/java/fk/prof/userapi/api/cache/ZkLoadInfoStore.java

```diff
@@ -0,0 +1,312 @@
+                .and().delete().forPath(zkPathForProfile(profileName))
+                .and().commit();
+        }
+        else {
+            zookeeper.setData().forPath(zkNodesInfoPath, newData);
+        }
+    }
+
+    AutoCloseable getLock() throws Exception {
+        return getLock(false);
+    }
+
+    AutoCloseable getLock(Boolean canExpectAlreadyAcquired) throws Exception {
+        ensureConnected();
+        return new CloseableSharedLock(canExpectAlreadyAcquired);
+    }
+
+    ConnectionState getState() {
+        return connectionState.get();
+    }
+
+    public enum ConnectionState {
+        Disconnected,
+        Connected
+    }
+
+    private void ensureConnected() throws ZkStoreUnavailableException {
+        if(!ConnectionState.Connected.equals(connectionState.get())) {
+            throw new ZkStoreUnavailableException("connection lost recently: " + recentlyZkConnectionLost.get() + ", lastTimeOfLostConnection: " + lastZkLostTime.get().toString());
+        }
+    }
+
+    private boolean pathExists(String path) throws  Exception {
+        return zookeeper.checkExists().forPath(path) != null;
+    }
+
+    private class CloseableSharedLock implements AutoCloseable {
+
+        boolean lockAlreadyAcquired = false;
+
+        CloseableSharedLock(boolean canExpectAlreadyAcquired) throws Exception {
+            if (!canExpectAlreadyAcquired && sharedMutex.isAcquiredInThisProcess()) {
+                throw new RuntimeException("Lock already acquired");
+            }
+            if (sharedMutex.isAcquiredInThisProcess()) {
+                lockAlreadyAcquired = true;
+            }
+            else {
+                if(!sharedMutex.acquire(5, TimeUnit.SECONDS)) {
+                    lockAcquireTimeoutCounter.inc();
+                    throw new ZkStoreUnavailableException("Could not acquire lock in the given time.");
+                }
+            }
+        }
+
+        @Override
+        public void close() {
+            try {
+                if (!lockAlreadyAcquired) {
+                    sharedMutex.release();
+                }
+            }
+            catch (Exception e) {
+                logger.error("Unable to release lock", e);
+            }
+        }
+    }
+
+    private NodeLoadInfo buildNodeLoadInfo(int profileCount) {
+        return NodeLoadInfo.newBuilder().setProfilesLoaded(profileCount).build();
+    }
+
+    private <T extends AbstractMessage> T readFrom(String path, Parser<T> parser) throws Exception {
+        byte[] bytes = zookeeper.getData().forPath(path);
+        return parser.parseFrom(bytes);
+    }
+
+
+    private String zkPathForProfile(AggregatedProfileNamingStrategy profileName) {
+        return "/profilesLoadStatus/" +  BaseEncoding.base32().encode(profileName.toString().getBytes(Charset.forName("utf-8")));
+    }
+
+    private void zkStateChangeListener(CuratorFramework client, org.apache.curator.framework.state.ConnectionState newState) {
```

Comment:
We have decided to handle zk connection state thusly:
1. Close the gate if LOST/SUSPENDED received.
2. If RECONNECT received: check ownership of ephemeral node, if not then clear local cache, setup ephemeral node => basically treat this as a fresh invocation of zkstore. Might want to abstract this in a init() method which is called in constructor as well as reinit
Open the gate if reinit succeeds
3. All public facing methods of zk store go via the gate. Important to ensure no race with reinit operations